### ISSUE TITLE
USB Keyboard - Azerty - german - french layout - how to ?

### ISSUE THREAD

**Issue by Olivier-ADLER** [k3ng/k3ng_cw_keyer#126]
I did not find any information about the method to include a different USB keyboard layout. It seems that only Qwerty US is supported.

Could someone give me the method ?

**Comment by Olivier-ADLER** [comment]
Nobody knows ?

**Comment by lengel67** [comment]
Hello, 
you can change the layout of the keyboard : https://docs.arduino.cc/language-reference/en/functions/usb/Keyboard/keyboardBegin/
In this case, you can modify the code to this : 
```arduino
#ifdef FEATURE_CW_COMPUTER_KEYBOARD
  Keyboard.begin(KeyboardLayout_fr_FR);
#endif //FEATURE_CW_COMPUTER_KEYBOARD
```

**Comment by Olivier-ADLER** [comment]
Thanks but i think that this is for keyboard emulation, to output keyboard keycodes.

What i need is HID USB keyboard input with FR layout.

**Comment by lengel67** [comment]
> Thanks but i think that this is for keyboard emulation, to output keyboard keycodes.

Yes indeed. Now, I see your problem but I don't have an idea on how to solve this, sorry.

**Comment by Olivier-ADLER** [comment]
After looking for a solution, i've found that the keyboard layout is defined into the hidboot.cpp file, inside the USB_Host_Shield library.

The code seems to be located in the OemToAscii() function:

`uint8_t KeyboardReportParser::OemToAscii(uint8_t mod, uint8_t key) {
        uint8_t shift = (mod & 0x22);

        // [a-z]
        if (VALUE_WITHIN(key, 0x04, 0x1d)) {
                // Upper case letters
                if ((kbdLockingKeys.kbdLeds.bmCapsLock == 0 && shift) ||
                        (kbdLockingKeys.kbdLeds.bmCapsLock == 1 && shift == 0))
                        return (key - 4 + 'A');

                        // Lower case letters
                else
                        return (key - 4 + 'a');
        }// Numbers
        else if (VALUE_WITHIN(key, 0x1e, 0x27)) {
                if (shift)
                        return ((uint8_t)pgm_read_byte(&getNumKeys()[key - 0x1e]));
                else
                        return ((key == UHS_HID_BOOT_KEY_ZERO) ? '0' : key - 0x1e + '1');
        }// Keypad Numbers
        else if(VALUE_WITHIN(key, 0x59, 0x61)) {
                if(kbdLockingKeys.kbdLeds.bmNumLock == 1)
                        return (key - 0x59 + '1');
        } else if(VALUE_WITHIN(key, 0x2d, 0x38))
                return ((shift) ? (uint8_t)pgm_read_byte(&getSymKeysUp()[key - 0x2d]) : (uint8_t)pgm_read_byte(&getSymKeysLo()[key - 0x2d]));
        else if(VALUE_WITHIN(key, 0x54, 0x58))
                return (uint8_t)pgm_read_byte(&getPadKeys()[key - 0x54]);
        else {
                switch(key) {
                        case UHS_HID_BOOT_KEY_SPACE: return (0x20);
                        case UHS_HID_BOOT_KEY_ENTER: return (0x13);
                        case UHS_HID_BOOT_KEY_ZERO2: return ((kbdLockingKeys.kbdLeds.bmNumLock == 1) ? '0': 0);
                        case UHS_HID_BOOT_KEY_PERIOD: return ((kbdLockingKeys.kbdLeds.bmNumLock == 1) ? '.': 0);
                }
        }
        return ( 0);
}
`
Then for letters, replacing by this for example:

`if (VALUE_WITHIN(key, 0x04, 0x1d)) {
    // Upper case letters
    if ((kbdLockingKeys.kbdLeds.bmCapsLock == 0 && shift) ||
        (kbdLockingKeys.kbdLeds.bmCapsLock == 1 && shift == 0))
        return AZERTY_uppercase_map[key - 0x04];

    // Lower case letters
    else
        return AZERTY_lowercase_map[key - 0x04];
}
const char AZERTY_lowercase_map[] = {
  'q', 'b', 'c', 'd', 'e', 'f', 'g', 'h',
  'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p',
  'a', 'r', 's', 't', 'u', 'v', 'z', 'x',
  'y', 'w'
};

const char AZERTY_uppercase_map[] = {
  'Q', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
  'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
  'A', 'R', 'S', 'T', 'U', 'V', 'Z', 'X',
  'Y', 'W'
};
`
